src/JuMP.jl

```diff
@@ -335,21 +335,126 @@ Base.broadcastable(model::Model) = Ref(model)
+
+If `attach_caching_optimizers`, `MOIU.attach_optimizer(model)` will be called on
+all `MOIU.CachingOptimizer`s in the stack.
+
+**This function should only be used by advanced users looking to access
+low-level solver-specific functionality. It has a high-risk of incorrect usage.
+We strongly suggest you use the alternative suggested below.**
+
+See also: [`backend`](@ref).
+
+## Unsafe behavior
+
+This function is unsafe because the innermost optimizer may not represent the
+same problem as `model`.
+
+If you modify the JuMP model, the reference you have to the backend may be
+invalid. Using an invalid reference may result in incorrect solutions, errors,
+or Julia crashing. Always get a new reference by calling `unsafe_backend` before
+calling any low-level functionality.
+
+Moreover, if you modify the unsafe backend, e.g., by adding a new constraint,
+the changes may be silently discarded by JuMP when the model is modified or
+solved. So don't modify the unsafe backend unless you understand when and how
+JuMP attaches and detaches optimizers! Use the alternative suggested below.
+
+If you pass `attach_caching_optimizers = false`, you should assume that any
+`CachingOptimizer`s may be in the `EMPTY_OPTIMIZER` state.
+
+## Alternative
+
+Instead of `unsafe_backend`, create a model using [`direct_model`](@ref) and
+call [`backend`](@ref) instead.
+
+For example, instead of:
+```julia
+model = Model(GLPK.Optimizer)
+@variable(model, x >= 0)
+glpk = unsafe_backend(model)
+# ... use GLPK ...
+@objective(model, Min, 2x + 1)
+glpk = unsafe_backend(model)
+# ... use GLPK again...
+```
+Use;
+```julia
+model = direct_model(GLPK.Optimizer())
+@variable(model, x >= 0)
+glpk = backend(model)
+# ... use GLPK ...
+@objective(model, Min, 2x + 1)
+# ... use GLPK again. No need to call `backend` twice...
+```
+"""
+function unsafe_backend(model::Model; attach_caching_optimizers::Bool = true)
+    return unsafe_backend(
+        backend(model);
+        attach_caching_optimizers = attach_caching_optimizers,
+    )
+end
+
+function unsafe_backend(
+    model::MOIU.CachingOptimizer;
+    attach_caching_optimizers::Bool,
+)
+    if MOIU.state(model) == MOIU.NO_OPTIMIZER
+        error(
+            "Unable to get backend optimizer because CachingOptimizer is " *
+            "in state `NO_OPTIMIZER`. Call [`set_optimizer`](@ref) first.",
+        )
+    elseif attach_caching_optimizers &&
+           MOIU.state(model) == MOIU.EMPTY_OPTIMIZER
+        MOIU.attach_optimizer(model)
+    end
+    return unsafe_backend(
+        model.optimizer;
+        attach_caching_optimizers = attach_caching_optimizers,
+    )
+end
+
+function unsafe_backend(
+    model::MOIB.LazyBridgeOptimizer;
+    attach_caching_optimizers::Bool,
+)
+    return unsafe_backend(
+        model.model;
+        attach_caching_optimizers = attach_caching_optimizers,
+    )
+end
+
+unsafe_backend(model::MOI.ModelLike; kwargs...) = model
```

Comment:
Are we recommending extensions and layers like Dualization.jl do something?

Reply:
They could, but that would require them to have JuMP as a dependency. Getting the backend through a dualize model seems tricky.